### DISCUSSION TITLE
Whether to support hypergraph neural networks, such as HGNN

### DISCUSSION THREAD

**Discussion by yaoysyao** [pyg-team/pytorch_geometric#7120] [Q&A] [answered]
Whether to support hypergraph neural networks, such as HGNN(https://arxiv.org/abs/1809.09401)

**Comment by wsad1** [comment]
We have a [hypergraphconv](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.HypergraphConv.html#torch_geometric.nn.conv.HypergraphConv), do you think we can use a similar input representation to implement the HGNN?
It would be nice to support this. Happy to accept contributions.

**Reply by yaoysyao** [reply]
Thank you for your reply and thank you for your contribution to PYG，I think it is more difficult in hypergraph neural networks to construct hypergraphs to facilitate model training, if pyg has a related method of constructing hypergraphs, I think you can try to use pyg to construct hypergraphs to complete tasks such as classification, now hypergraph neural networks have been used by more and more researchers, how to simply construct hypergraph neural networks is a problem for many researchers

**Reply by wsad1** [reply]
Right. So to support hypergraphs better in pyg, we'll might need the following
1. `HyperGraphData` which will mostly inherit from `Data` and override some methods.
2. A `HyperGraphLoader` that can create minibatches of `HyperGraphData` from a list.
3. New HyperGraph GNNs.
4. Several example Datasets of `HyperGraphData`.
5. Maybe even a `HyperGraphNeighborLoader`.

These are just some initial thoughts open to suggestions.
cc: @rusty1s.

**Reply by yaoysyao** [reply]
> Right. So to support hypergraphs better in pyg, we'll might need the following
> 
> 1. `HyperGraphData` which will mostly inherit from `Data` and override some methods.
> 2. A `HyperGraphLoader` that can create minibatches of `HyperGraphData` from a list.
> 3. New HyperGraph GNNs.
> 4. Several example Datasets of `HyperGraphData`.
> 5. Maybe even a `HyperGraphNeighborLoader`.
> 
> These are just some initial thoughts open to suggestions. cc: @rusty1s.

Thanks for the reply, you are quite right，but I think not only supergraph GNNS, due to the current use of attention mechanism more and more frequently, may also need hypergraph attention neural network, and because more and more researchers begin to use graph neural networks to solve NLP-related problems, you can try to support NLP-related tasks

**Reply by yaoysyao** [reply]
In addition, I have another question I want to ask, whether the dataset in the PYG includes a dataset of text classification, I have checked the relevant source code, it seems that I have not found a dataset for text classification, can I provide a related dataset, or can I provide a relevant demo to facilitate me to create my own dataset to solve the text classification task

**Comment by rusty1s** [comment, marked as answer]
@wsad1 @yaoysyao I am definitely open for better hypergraph support, but IMO this is already well supported by `DataLoader` as long as you use a customized `Data` variant that increases `hyperedge_index` according to the shape of the underlying incidence matrix.

**Reply by yaoysyao** [reply]
Thank you for your reply, thank you for creating such a good tool, according to your reply, do you mean that as long as I can construct a graph structure that conforms to the PYG format, even if it is a hypergraph, I can use PYG for graph classification or node classification?

**Reply by rusty1s** [reply]
Yes, that is correct. You can use `HyperGraphConv` for all of these tasks. I am happy to include more hyperedge GNN layers in the future.
